core/trino-main/src/main/java/io/trino/operator/exchange/UniformPartitionRebalancer.java

```diff
@@ -0,0 +1,430 @@
+    {
+        PartitionInfo partitionInfo = partitionInfos[partitionId];
+        int writerIndex = floorMod(index, partitionInfo.getWriterCount());
+        return partitionInfo.getWriter(writerIndex);
+    }
+
+    @VisibleForTesting
+    List<Integer> getWriterIds(int partitionId)
+    {
+        return partitionInfos[partitionId].getWriterIds();
+    }
+
+    public void rebalancePartitions()
+    {
+        List<Long> writerPhysicalWrittenBytes = writerPhysicalWrittenBytesSuppliers.stream()
+                .map(Supplier::get)
+                .collect(toImmutableList());
+
+        // Rebalance only when total bytes written since last rebalance is greater than MIN_PHYSICAL_WRITTEN_REBALANCE_BYTES (100 MB)
+        if (getPhysicalWrittenBytesSinceLastRebalance(writerPhysicalWrittenBytes) > REBALANCE_THRESHOLD_MIN_PHYSICAL_WRITTEN_BYTES) {
+            rebalancePartitions(writerPhysicalWrittenBytes);
+        }
+    }
+
+    private int getPhysicalWrittenBytesSinceLastRebalance(List<Long> writerPhysicalWrittenBytes)
+    {
+        int physicalWrittenBytesSinceLastRebalance = 0;
+        for (int writerId = 0; writerId < writerPhysicalWrittenBytes.size(); writerId++) {
+            physicalWrittenBytesSinceLastRebalance +=
+                    writerPhysicalWrittenBytes.get(writerId) - writerPhysicalWrittenBytesAtLastRebalance.get(writerId);
+        }
+
+        return physicalWrittenBytesSinceLastRebalance;
+    }
+
+    private synchronized void rebalancePartitions(List<Long> writerPhysicalWrittenBytes)
+    {
+        Long2LongMap partitionRowCounts = partitionRowCountsSupplier.get();
+        RebalanceContext context = new RebalanceContext(writerPhysicalWrittenBytes, partitionRowCounts);
+
+        IndexedPriorityQueue<WriterId> maxWriters = new IndexedPriorityQueue<>();
+        IndexedPriorityQueue<WriterId> minWriters = new IndexedPriorityQueue<>();
+        for (int writerId = 0; writerId < numberOfWriters; writerId++) {
+            WriterId writer = new WriterId(writerId);
+            maxWriters.addOrUpdate(writer, context.getWriterPriority(writer));
+            minWriters.addOrUpdate(writer, Long.MAX_VALUE - context.getWriterPriority(writer));
+        }
+
+        // Find skewed partitions and scale them across multiple writers
+        while (true) {
+            // Find the writer with maximum physical written bytes since last rebalance
+            WriterId maxWriter = maxWriters.poll();
+
+            if (maxWriter == null) {
+                break;
+            }
+
+            // Find the skewness against writer with max physical written bytes since last rebalance
+            List<WriterId> minSkewedWriters = findSkewedMinWriters(context, maxWriter, minWriters);
+            if (minSkewedWriters.isEmpty()) {
+                break;
+            }
+
+            for (WriterId minSkewedWriter : minSkewedWriters) {
+                if (context.rebalancePartition(maxWriter, minSkewedWriter)) {
```

Comment:
Can it happen that no `rebalancePartition` will be successful? If so, you most likely should quit `while (true) {` loop, otherwise it might loop forever